ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,421 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+    NT_ASSERT(NT_SUCCESS(Status));
+
+    if (ReferenceString != NULL)
+    {
+        Status = RtlAppendUnicodeStringToString(&SymbolicLink, ReferenceString);
+        NT_ASSERT(NT_SUCCESS(Status));
+    }
     else
-        DeviceKeyRealP = &DeviceKeyReal;
+    {
+        Status = RtlAppendUnicodeToString(&SymbolicLink, L"\\");
+        NT_ASSERT(NT_SUCCESS(Status));
+    }
+
+    DPRINT("New symbolic link is '%wZ'\n", &SymbolicLink);
+
+    *SymbolicLinkName = SymbolicLink;
+    return STATUS_SUCCESS;
+}
+
+static
+NTSTATUS
+IopSeparateSymbolicLink(_In_ PCUNICODE_STRING SymbolicLinkName,
+                        _Out_opt_ PUNICODE_STRING PrefixString,
+                        _Out_opt_ PUNICODE_STRING MungedString,
+                        _Out_opt_ PUNICODE_STRING GuidString,
+                        _Out_opt_ PUNICODE_STRING ReferenceString,
+                        _Out_opt_ PBOOLEAN ReferenceStringPresent,
+                        _Out_opt_ LPGUID InterfaceClassGuid)
+{
+    UNICODE_STRING KernelModePrefix = RTL_CONSTANT_STRING(L"\\??\\");
+    UNICODE_STRING UserModePrefix = RTL_CONSTANT_STRING(L"\\\\?\\");
+    UNICODE_STRING MungedPrefix = RTL_CONSTANT_STRING(L"##?#");
+    UNICODE_STRING MungedStringReal, GuidStringReal, ReferenceStringReal;
+    UNICODE_STRING DeviceString;
+    UNICODE_STRING LinkNameNoPrefix;
+    USHORT i, ReferenceStringOffset;
+    NTSTATUS Status;
+
+    MungedStringReal.Buffer = GuidStringReal.Buffer = ReferenceStringReal.Buffer = NULL;
+
+    DPRINT("Symbolic link is '%wZ'\n", SymbolicLinkName);
+
+    /* Symbolic link name looks like \??\ACPI#PNP0501#1#{GUID}\ReferenceString */
+    /* Make sure it starts with the expected prefix */
+    if (!RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE) &&
+        !RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE))
+    {
+        DPRINT1("Invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Sanity checks */
+    ASSERT(KernelModePrefix.Length == UserModePrefix.Length);
+    ASSERT(SymbolicLinkName->Length >= KernelModePrefix.Length);
 
-    if (InstanceKey != NULL)
-        InstanceKeyRealP = InstanceKey;
+    /* Make a version without the prefix for further processing */
+    LinkNameNoPrefix.Buffer = SymbolicLinkName->Buffer + KernelModePrefix.Length / sizeof(WCHAR);
+    LinkNameNoPrefix.Length = SymbolicLinkName->Length - KernelModePrefix.Length;
+    LinkNameNoPrefix.MaximumLength = LinkNameNoPrefix.Length;
+
+    DPRINT("Symbolic link without prefix is '%wZ'\n", &LinkNameNoPrefix);
+
+    /* Find the reference string, if any */
+    for (i = 0; i < LinkNameNoPrefix.Length / sizeof(WCHAR); i++)
+    {
+        if (LinkNameNoPrefix.Buffer[i] == L'\\')
+        {
+            break;
+        }
+    }
+    ReferenceStringOffset = i * sizeof(WCHAR);
+
+    /* The GUID is before the reference string or at the end */
+    ASSERT(SymbolicLinkName->Length >= ReferenceStringOffset);
```

Comment:
I think you want `LinkNameNoPrefix->Length` here instead of `SymbolicLinkName`'s, since `ReferenceStringOffset` is relative to `LinkNameNoPrefix`

Reply:
OK, fixed.